ocaml/typing/persistent_env.ml

```diff
@@ -344,23 +397,280 @@ let find_import ~allow_hidden penv ~check modname =
+    in
+    (* Only need to substitute the visible args, since the hidden args only
+       reflect substitutions already made by visible args *)
+    Signature_with_global_bindings.subst sign bindings
+  in
+  Array.iter
+    (fun bound_global ->
+       remember_global penv bound_global ~mentioned_by:global_name)
+    sign.bound_globals;
+  let pn = { pn_import = import;
+             pn_global = global;
+             pn_arg_for = arg_for;
+             pn_sign = sign.sign;
+           } in
+  if check then check_consistency penv import;
+  Hashtbl.add persistent_names global_name pn;
+  remember_global penv global ~mentioned_by:global_name;
+  pn
+
+and find_pers_name ~allow_hidden penv check name =
+  let {persistent_names; _} = penv in
+  match Hashtbl.find persistent_names name with
+  | pn -> pn
+  | exception Not_found ->
+      let unit_name = CU.Name.of_head_of_global_name name in
+      let import = find_import ~allow_hidden penv ~check unit_name in
+      acknowledge_pers_name penv check name import
+
+let read_pers_name penv check name filename =
+  let unit_name = CU.Name.of_head_of_global_name name in
+  let import = read_import penv ~check unit_name filename in
+  acknowledge_pers_name penv check name import
+
+let need_local_ident penv (global : Global_module.t) =
+  (* There are three equivalent ways to phrase the question we're asking here:
+
+     1. Is this either a parameter or an open import?
+     2. Will the generated lambda code need a parameter to take this module's
+          value?
+     3. Is the value not statically bound?
+
+     Crucially, all modules (besides the one being compiled or instantiated)
+     must be either statically bound or toplevel parameters, since the actual
+     functor calls that instantiate open modules happen elsewhere (so that they
+     can happen exactly once). *)
+  let global_name = global |> Global_module.to_name in
+  let name = global_name |> CU.Name.of_head_of_global_name in
+  if is_registered_parameter_import penv global_name
+  then
+    (* Already a parameter *)
+    true
+  else if current_unit_is name
+  then
+    (* Not actually importing it in the sense of needing its value (we're
+       building its value!) *)
+    false
+  else if current_unit_is_instance_of name
+  then
+    (* We're instantiating the module, so (here and only here!) we're accessing
+       its actual functor, which is a compile-time constant *)
+    (* CR lmaurer: Relying on [current_unit_is_instance_of] here feels hacky
+       when only a pretty specific call sequence gets here. *)
+    false
+  else if Global_module.is_complete global
+  then
+    (* It's a compile-time constant *)
+    false
+  else
+    (* Some argument is missing, or some argument's argument is missing, etc.,
+       so it's not a compile-time constant *)
+    true
+
+let make_binding penv (global : Global_module.t) (impl : CU.t option) : binding =
+  let name = Global_module.to_name global in
```

Comment:
I'm not sure how that helps? `compute_global` is about filling in hidden information in `global` and `make_binding` uses that information. That's what's so useful about `Global_module.t`: it makes things explicit so that, for instance, you can look at it and figure out whether it requires a local binding.

Also, I'm not sure what you mean about having access to both the `t` and the `Name.t`? If you have access to a `Global_module.t`, you already have access to the corresponding `Global_module.Name.t`. So if you need to call `is_registered_parameter_import`, which takes a `Global_module.Name.t` (and here it really does make sense that it might or might not have arguments), you're not sad that you have a `Global_module.t` instead. You just project.

Reply:
Yes after getting my hands dirty, I agree that `need_local_ident` can be delayed as needed, because `Global_module.t` already contains much information.

However, I'm still not happy with the fact that `Global_module.t` doesn't encode the info about whether it's a parameter. 
I think to follow the intuition that "Global_module.t is resolved reference` (as described in https://github.com/ocaml-flambda/flambda-backend/pull/1726#discussion_r1812785870), it should be defined as:
```
type t = 
  | Parameter of Parameter.t
  | Normal of {head : string; visible_args; hidden_args}
```
And `compute_global` should check that, if some `Name.t` is a parameter, is registered as such. (it's similar to checking some `Longident.t` is bound and returns a `Ident.t`)

Reply:
The intention isn't really to remember everything that we learn through resolution, it's to encode enough information that substitution can operate syntactically. Most things that we learn when resolving an identifier don't go into the resolved identifier, they go into the tables in `Env` and `Persistent_env`, and I'd prefer to stay consistent with that when practical.

Reply:
Yes that makes sense. Let's explore this further in the future PR where `ident.t` uses `Global_module.t`. I'm resolving this for now.